---
title: "DADA2 Workflow for BIOMI6300 Project 2"
author: "Luna Eresta Jaya"
date: "2023-04-26"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Loading packages
```{r load}
# Set working directory
setwd("/workdir/lej52/BIOMI6300_Project2_Amplicon_Analysis")

# Load packman packages
# install.packages("pacman")
# Download Software 
pacman::p_load(dada2, tidyverse, patchwork, phyloseq, Biostrings, install = FALSE)
# Ensure correct package
packageVersion("dada2")

# Load in the functions file from code directory
source("/local/workdir/lej52/BIOMI6300_Project2_Amplicon_Analysis/code/functions.R")
```

# Set the path to the seq files
```{r}
# Set path to the gzipped files
path <- "data/sequencing_3"
path

# What files do we have?
list.files(path)

# Use 4 samples first to test workflow
samples <- c("JD-11", "JD-217", "JD-111", "JD-110")
```

# Load in Forward and Reverse reads and assess the quality
```{r}
# Load reads
# Create variable for the forward and the reverse reads

# Forward read variable
forward_reads <- sort(list.files(path, pattern="_1.fastq",
                      full.names=TRUE))

# Reverse read variable
reverse_reads <- sort(list.files(path, pattern="_2.fastq",
                      full.names=TRUE))

# 3. Place filtered files into filtered/subdirectory
# Create a variable holding file names for the Forward and Reverse filtered reads
filtered_forward_reads <- file.path(path, "filtered", paste0(samples, "_R1_filtered.fastq.gz"))
filtered_reverse_reads <- file.path(path, "filtered", paste0(samples, "_R2_filtered.fastq.gz"))

# Show the quality of each base on the reads of first 4 samples
forwardQual4_plot <- plotQualityProfile(forward_reads[1:4])
reverseQual4_plot <- plotQualityProfile(reverse_reads[1:4])

# Plot forward and reverse 4 quality plots together
forwardQual4_plot+reverseQual4_plot
```

# Filtering and trimming
```{r}
filtered_out <- filterAndTrim(forward_reads, filtered_forward_reads,
                              reverse_reads, filtered_reverse_reads,
                              truncLen = c(100, 100), trimLeft = c(20,20),
                              maxN = 0, maxEE = c(1,1), truncQ = 2,
                              rm.phix = TRUE, compress = TRUE,
                              multithread = TRUE)

# Use figaro by Zymo-Research (on github) to trim auotomatically